### ISSUE TITLE
Tree view reload after assembly is reloaded

### ISSUE THREAD

**Issue by MartinZec** [TestCentric/testcentric-gui#887]
seen at 2.0.0-alpha5: using the Automatic Reload feature: Reload before each test run.

I used a big test library, expanding a long tree, with 6 levels. When I execute a test, first the assembly is reloaded. By this, the tree is reloaded but only until the 4th level, so I cannot see the executed test any more.

Also, only the first node is expanded after reloading. In case a test of, for example, the fifth node is executed, after reloading still the first node is expanded after the tree reloads.

**Comment by CharliePoole** [comment]
The problem seems to exist whether the reload is automatic or manual. I believe it comes down to the use of the saved `VisualState` behind the scenes, to restore the tree to it's appearance before the reload. I've seen other `VisualState` glitches from time to time, so I think it needs a closer look before the next release.

Do you have "Save Visual State of each project" on or off in your Tree Settings? I get different results (both bad, but different) in each case. When it's off, I get the default tree expansion. When it's on I get something else - apparently a previously saved state - but not the state the tree was in when I initiated the reload.

**Comment by MartinZec** [comment]
I had "Save Visual State of each project" enabled, as it is by default. Disabling it make it a bit better, but still not as it should be in my point of view.

**Comment by CharliePoole** [comment]
For testing purposes, it's better to have it disabled. That's because the internal use of save + restore during reload is entirely in memory. After the reload completes, whatever visual state was last saved in the file at application close is being reloaded. That's actually wrong as well, but it's a separate problem in the code. Right now I'm just adding more tests to ensure that save and restore work correctly by themselves, independent of how they are called.

**Comment by CharliePoole** [comment]
This relates to #774 as well. Visual State is only handled for the NUnit Tree display strategy. In fact , that's where the code to save and restore it lives! That's because... history. :-)

In order to fix this issue properly, I'll have to fix that one as well.

**Comment by CharliePoole** [comment]
PR #888 fixes some bugs in VisualState itself and puts all it's operation in the class itself. Previously, the class didn't know how to apply itself to a  tree - now it does. As a result, the new build (2.0.0-dev-00135) correctly restores the visual state when an assembly is re-opened, provided no changes have been made in the interim. That's actually the main purpose of VisualState, so it's a win.

However, this doesn't fix the current issue. It just narrows it down to problems with how Reload __makes use of__ VisualState.

NOTE: Because the current VisualState algorithm relies on the test ids provided by the NUnit framework, it can't handle changes to the tests if those changes result in changing the ids. For that reason, I'm limiting this issue to reloading the tests __when no such changes have been made.__ That's obviously a pretty big limitation, but if we can't even do a proper reload when nothing has changed, it will be nearly impossible to deal with actual changes to the tests. Once this issue is complete, I'll be creating other issues to deal with each type of change, like renaming a test, removing a test, etc.

**Comment by CharliePoole** [comment]
Moving forward... tests are reloaded in the following situations...

1. On command by the user
2. When tests are run if "Reload before each test run" is set
3. When tests are re-compiled if "Reload on change" is set
4. When the display strategy or grouping changes

In some cases, the reload may encounter changes in the tests not currently reflected in the display. We are not dealing with those changes in this issue, for reasons stated in the preceding comment.

This issue was written to deal with item 2 in the list. However, I'll attempt to deal with the first three items here because I think the same underlying code is used. The fourth item is somewhat moot, since we don't yet save visual state unless the NUnit tree strategy is in use.

**Comment by CharliePoole** [comment]
I believe this is fixed for both cases 1 and 2. The fix is available on the MyGet feed as version 2.0.0-dev-00137.

@MartinZec Let me know if you see any problem in the scope of this issue. I'll write another issue for restoring the visual state after various kinds of changes, since that's a bigger can of worms.

**Comment by CharliePoole** [comment]
@MartinZec That's a good point. It only makes sense if one understands the internal structure of the application, which isn't a good thing at all! So I agree that in the case of a reload, that setting shouldn't apply - we should always save and restore the visual state.

I can see three options here, all of them pretty easy...

1. Just do what we are now doing whether the setting is checked or not.
2. Only create a new saved file if the option is checked, otherwise save the visual state in memory before reloading and restore it after.
3. Eliminate the option entirely and save / use visual state at all times, with the possible exception of running the application unattended. I actually like this one myself unless we think of some use case that it breaks.

Could you create a new issue for this and include something about which option you prefer?

**Comment by CharliePoole** [comment]
@MartinZec Also... if you'd like to do the PR, you're welcome to do so. :-)
